### PR TITLE
JDK-8302102: Disable ASan for SafeFetch and os::print_hex_dump

### PR DESCRIPTION
**Pull request by jcking**
Disable ASan instrumentation for non-assembly versions of SafeFetch and os::print_hex_dump.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302102](https://bugs.openjdk.org/browse/JDK-8302102): Disable ASan for SafeFetch and os::print_hex_dump

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [a4524758](https://git.openjdk.org/jdk/pull/12477/files/a452475849d1674e58b88f46914cdbfec3089939)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12477/head:pull/12477` \
`$ git checkout pull/12477`

Update a local copy of the PR: \
`$ git checkout pull/12477` \
`$ git pull https://git.openjdk.org/jdk pull/12477/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12477`

View PR using the GUI difftool: \
`$ git pr show -t 12477`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12477.diff">https://git.openjdk.org/jdk/pull/12477.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Seems fine. Thanks.

**Review by tstuefe — APPROVED** [review]
Seems fine. 

Please build and test zero before pushing (gtests are fine).

**Comment by jcking** [comment]
> Seems fine.
> 
> Please build and test zero before pushing (gtests are fine).

If we are talking about without ASan, it works fine. With ASan is a different story...

So, Zero does a bunch of things that are not acceptable and fails to finish the build while optimizing the image. It triggers stack-use-after-scope and global-buffer-overflow. The first is in `ZeroInterpreter::native_entry(Method*, long, JavaThread*) src/hotspot/cpu/zero/zeroInterpreter_zero.cpp:435` because it looks like it takes the address of `env` above in a block and then leaves the block, leaving the address of `env` undefined. The second is related to Atomics `unsigned int Atomic::PlatformLoad<4ul>::operator()<unsigned int>(unsigned int const volatile*) const src/hotspot/share/runtime/atomic.hpp:432` because Zero performs a 4-byte load for a variable with 1-byte storage.

I am not super interested in fixing Zero, though these issues should definitely be reported and fixed. They are all undefined behavior and likely corruption in the first case. Shall I report them and proceed with integration?

File: src/hotspot/os/windows/safefetch_windows.hpp [hunk 1]
```diff
@@ -26,12 +26,13 @@
...
 #define OS_WINDOWS_SAFEFETCH_WINDOWS_HPP
 
+#include "sanitizers/address.hpp"
 #include "utilities/globalDefinitions.hpp"
 
 // On windows, we use structured exception handling to implement SafeFetch
 
 template <class T>
-inline T SafeFetchXX(const T* adr, T errValue) {
+ATTRIBUTE_NO_SANITIZE_ADDRESS static inline T SafeFetchXX(const T* adr, T errValue) {
```
Comment (tstuefe) [inline src/hotspot/os/windows/safefetch_windows.hpp:35]: Wait, why is this static now?

File: src/hotspot/share/sanitizers/address.hpp [hunk 2]
```diff
@@ -29,6 +29,25 @@
 #include <sanitizer/asan_interface.h>
 #endif
 
+// ATTRIBUTE_NO_SANITIZE_ADDRESS
```
Comment (tstuefe) [inline src/hotspot/share/sanitizers/address.hpp:32]: Can we find a better name for this? Its lengthy but still not very clear. E.g. as a prefix to a function it reads like it does something to the - should there be one - return pointer of the function.

proposal: ASAN_DISABLE

**Comment by tstuefe** [comment]
> I am not super interested in fixing Zero, though these issues should definitely be reported and fixed. They are all undefined behavior and likely corruption in the first case. Shall I report them and proceed with integration?

I was assuming you were concerned with zero because you changed the sigjmp variant of SafeFetch, which is only used on zero.

File: src/hotspot/os/windows/safefetch_windows.hpp (hunk 1 above)
Comment (jcking) [reply src/hotspot/os/windows/safefetch_windows.hpp]: Habit. Removed it to be consistent with the rest of the declarations.

File: src/hotspot/share/sanitizers/address.hpp (hunk 2 above)
Comment (jcking) [reply src/hotspot/share/sanitizers/address.hpp:32]: I was mostly going for being consistent with other uses in globalDefinitions. Most things are ATTRIBUTE_X where X is the actual attribute name.

Maybe ATTRIBUTE_NO_ASAN if we want something shorter?

**Comment by jcking** [comment]
> > I am not super interested in fixing Zero, though these issues should definitely be reported and fixed. They are all undefined behavior and likely corruption in the first case. Shall I report them and proceed with integration?
> 
> I was assuming you were concerned with zero because you changed the sigjmp variant of SafeFetch, which is only used on zero.

I was mostly just going for completeness in case somebody wanted to use ASan with Zero, since I knew it wouldn't work just like the Windows one. I'll file bugs for the Zero bugs to be looked into.

File: src/hotspot/share/sanitizers/address.hpp (hunk 2 above)
Comment (tstuefe) [reply src/hotspot/share/sanitizers/address.hpp:32]: Sounds good.

Comment (jcking) [reply src/hotspot/share/sanitizers/address.hpp]: Renamed.

**Comment by jcking** [comment]
Filed [JDK-8302347](https://bugs.openjdk.org/browse/JDK-8302347) and [JDK-8302348](https://bugs.openjdk.org/browse/JDK-8302348) for Zero. Not sure which component, so just placed in Hotspot for now.

I'll wait for GHA to finish before integrating.

**Review by dholmes-ora — APPROVED** [review]

**Pull request closed without merging**